types/node/fs.d.ts

```diff
@@ -871,7 +872,23 @@ declare module "fs" {
      * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
      * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
      */
-    function mkdir(path: PathLike, options: number | string | MakeDirectoryOptions | undefined | null, callback: NoParamCallback): void;
+    function mkdir(path: PathLike, options: MakeDirectoryOptions & { recursive: true }, callback: (err: NodeJS.ErrnoException | null, path: string) => void): void;
+
+    /**
+     * Asynchronous mkdir(2) - create a directory.
+     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
+     * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
+     * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
+     */
+    function mkdir(path: PathLike, options: number | string | (MakeDirectoryOptions & { recursive?: false; }) | null | undefined, callback: NoParamCallback): void;
+
+    /**
+     * Asynchronous mkdir(2) - create a directory.
+     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
+     * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
+     * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
+     */
+    function mkdir(path: PathLike, options: number | string | MakeDirectoryOptions | null | undefined, callback: (err: NodeJS.ErrnoException | null, path: string | undefined) => void): void;
```

Comment:
Doesn't this line render the specialised overloads mood?
A string CAN only ever be returned if `recursive` IF an `MakeDirectoryOptions` is passed for the options.
I think you can remove the type union here and reduce it to just `MakeDirectoryOptions`

Reply:
This is the fallback overload to ensure that we don’t break existing code, where the exact subtype of `options` can’t be determined.

This fallback overload also exists for `child_process.spawn`: <https://github.com/DefinitelyTyped/DefinitelyTyped/blob/45aba8f505f4e353ecef3d0b589fe968836918b7/types/node/child_process.d.ts#L258>

---

This is necessary because of <https://github.com/microsoft/TypeScript/issues/14107>.